### PR TITLE
CSHARP-4658: NullReferenceException when using certain projections.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/PartialEvaluator.cs [hunk 1]
```diff
@@ -190,21 +190,17 @@ public override Expression Visit(Expression expression)
...
-                            throw new InvalidOperationException($"Unexpected binding type: {binding.BindingType}.");
-                    }
+                if (_cannotBeEvaluated)
+                {
+                    // visit only the arguments if any MemberBindings cannot be partially evaluated
+                    Visit(node.NewExpression.Arguments);
+                }
+                else
+                {
+                    Visit(node.NewExpression);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/PartialEvaluator.cs:202]: It's possible our whole `PartialEvaluator` class needs a good going over.

This particular issue is probably more easily explained on Zoom.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/PartialEvaluator.cs [hunk 2]
```diff
@@ -190,21 +190,17 @@ public override Expression Visit(Expression expression)
             protected override Expression VisitMemberInit(MemberInitExpression node)
             {
                 // Bindings must be visited before NewExpression
-                foreach (var binding in node.Bindings)
-                {
-                    switch (binding.BindingType)
-                    {
-                        case MemberBindingType.Assignment:
-                            var memberAssignment = (MemberAssignment)binding;
-                            base.Visit(memberAssignment.Expression);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/PartialEvaluator.cs:199]: Calling `base.Visit` (instead of `Visit`) is the core of the bug.

**Review by sanych-sun — APPROVED** [review]
LGTM

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**